crates/sui-framework/sources/publisher.move

```diff
@@ -0,0 +1,145 @@
+    /// a type originated from.
+    struct Publisher has key, store {
+        id: UID,
+        type: String
+    }
+
+    /// Claim a Publisher object.
+    /// Requires a One-Time-Witness to
+    public fun claim<OTW: drop>(otw: OTW, ctx: &mut TxContext): Publisher {
+        assert!(types::is_one_time_witness(&otw), ENotOneTimeWitness);
+
+        Publisher {
+            id: object::new(ctx),
+            type: type_name::into_string(type_name::get<OTW>()),
+        }
+    }
+
+    /// Claim a Publisher object and send it to transaction sender.
+    /// Since this function can only be called in the module initializer,
+    /// the sender is the publisher.
+    public fun claim_and_keep<OTW: drop>(otw: OTW, ctx: &mut TxContext) {
+        sui::transfer::transfer(claim(otw, ctx), sender(ctx))
+    }
+
+    /// Destroy a Publisher object effectively removing all privileges
+    /// associated with it.
+    public fun burn(publisher: Publisher) {
+        let Publisher { id, type: _ } = publisher;
+        object::delete(id);
+    }
+
+    /// Check whether type belongs to the same package as the publisher object.
+    public fun is_package<T>(publisher: &Publisher): bool {
+        let this = ascii::as_bytes(&publisher.type);
+        let their = ascii::as_bytes(type_name::borrow_string(&type_name::get<T>()));
+
+        let i = 0;
+
+        // 40 bytes => length of the HEX encoded string
+        while (i < 40) {
+            if (vector::borrow<u8>(this, i) != vector::borrow<u8>(their, i)) {
+                return false
+            };
+
+            i = i + 1;
+        };
+
+        true
+    }
+
+    /// Check whether a type belogs to the same module as the publisher object.
+    public fun is_module<T>(publisher: &Publisher): bool {
+        if (!is_package<T>(publisher)) {
+            return false
+        };
+
+        let this = ascii::as_bytes(&publisher.type);
+        let their = ascii::as_bytes(type_name::borrow_string(&type_name::get<T>()));
+
+        // 42 bytes => length of the HEX encoded string + :: (double colon)
+        let i = 42;
+        loop {
+            let left = vector::borrow<u8>(this, i);
+            let right = vector::borrow<u8>(their, i);
+
+            if (left == &ASCII_COLON && right == &ASCII_COLON) {
+                return true
+            };
+
+            if (left != right) {
+                return false
+            };
+
+            i = i + 1;
+        }
+    }
+
+    #[test_only]
+    /// Test-only function to claim a Publisher object bypassing OTW check.
+    public fun test_claim<OTW: drop>(_: OTW, ctx: &mut TxContext): Publisher {
+        Publisher {
+            id: object::new(ctx),
+            type: type_name::into_string(type_name::get<OTW>()),
+        }
+    }
+}
+
+#[test_only]
+module sui::test_publisher {
+    use sui::publisher;
+    use sui::test_scenario::{Self as test, Scenario, ctx};
+
+    /// OTW for the test_publisher module
+    struct TEST_OTW has drop {}
```

Comment:
How does his end up being interpreted as a OTW, given that its name doesn't match, is it some test-mode magic?  It would be good to document what's going on here, in case someone else needs to do the same thing.

Reply:
Haha. It's a workaround because I wasn't able to create an OTW in a `#[test]` function. If you look at how it's being used, there's a `test_claim` function instead of `claim` which allows any struct really as long as it has drop.